1.cellprofiler_ic/image_quality_control/scripts/1.evaluate_qc.py

```diff
@@ -0,0 +1,391 @@
+plt.xlabel("ImageQuality_PowerLogLogSlope")
+
+plt.tight_layout()
+
+plt.savefig(figure_dir / "blur_outliers_per_plate.png", dpi=500)
+
+plt.show()
+
+
+# ## Saturation metric
+# 
+# For saturation metrics, we are looking for:
+# 
+# - Smudged images or images containing large artifacts (usually seen in `DNA` channel regardless of stain)
+# - Overly saturated channels which can occur in any channel, but specifically we are seeing `Mito` having the most problems with this.
+# 
+# This means that we will only be setting threshold for outliers for all channels just like for blur.
+
+# In[10]:
+
+
+summary_statistics = df["ImageQuality_PercentMaximal"].describe()
+print(summary_statistics)
+
+
+# ### Visualize density plot of saturation metric
+
+# In[11]:
+
+
+# Create a KDE plot with separate lines for each Metadata_Plate
+plt.figure(figsize=(10, 6))
+sns.kdeplot(
+    x=np.log1p(df["ImageQuality_PercentMaximal"]), hue=df["Metadata_Plate"], fill=True
+)
+
+# Add a vertical line at the log-transformed value of 1.4
+plt.axvline(x=np.log1p(1.4), color='red', linestyle='--', label='Log-transformed 1.4')
+
+# Set labels and title
+plt.ylabel("Density")
+plt.xlabel("Log-transformed Percent Maximal")
+plt.title("KDE Plot of Log-transformed Percent Maximal by Plate")
+
+plt.savefig(figure_dir / "saturation_outliers_per_plate.png", dpi=500)
+
+# Show the plot
+plt.show()
+
+
+# ### Use z-score to find outliers 2 standard deviations above the mean to view as a data frame
+
+# In[12]:
+
+
+# Calculate Z-scores for the column
+z_scores = zscore(df["ImageQuality_PercentMaximal"])
+
+# Set a threshold for Z-scores (adjust as needed for number of standard deviations away from the mean)
+threshold_z = 2
+
+# Identify outlier rows based on Z-scores greater than as to identify whole images with abnormally high saturated pixels
+saturation_outliers = df[abs(z_scores) > threshold_z]
+
+# Identify unique well + site combos among outliers
+unique_combos = saturation_outliers[
+    ["Metadata_Well", "Metadata_Site"]
+].drop_duplicates()
+
+# Visualize example images through FIJI using data frame identifying outliers
+print(saturation_outliers["Channel"].value_counts())
+saturation_outliers.sort_values(by="ImageQuality_PercentMaximal", ascending=True).head()
+
+
+# ### Identify threshold values to use to find outliers above the mean
+# 
+# **Note:** This value will be used in CellProfiler to flag and not process images. We will round to the second decimal to use in the pipeline.
+
+# In[13]:
+
+
+# Calculate the mean and standard deviation
+mean_value = df["ImageQuality_PercentMaximal"].mean()
+std_dev = df["ImageQuality_PercentMaximal"].std()
+
+# Set the threshold multiplier for above the mean
+threshold = 2
+
+# Calculate the threshold values
+threshold_value_above_mean = mean_value + threshold * std_dev
+
+# Print the calculated threshold values
+print("Threshold for outliers above the mean:", threshold_value_above_mean)
```

Comment:
Are outliers only found above the mean using this metric (not also below the mean)?

Reply:
Correct, we are looking for images that are overly saturated as we can see that is what is most impacting the data. There are likely images that aren't saturated at all (empty and could be detected with PercentMinimal) but we expect these to be removed with blur metric.